mlrun/model.py

```diff
@@ -524,12 +539,142 @@ def is_hyper_job(self):
         return param_file or self.hyperparams
 
     @property
-    def inputs(self):
+    def inputs(self) -> Dict[str, str]:
+        """
+        Get the inputs dictionary. A dictionary of parameter names as keys and paths as values .
+
+        :return: The inputs dictionary.
+        """
         return self._inputs
 
     @inputs.setter
-    def inputs(self, inputs):
-        self._inputs = self._verify_dict(inputs, "inputs")
+    def inputs(self, inputs: Dict[str, str]):
+        """
+        Set the given inputs in the spec. Inputs can include a type hint string in their keys following a colon, meaning
+        following this structure: "<input key : type hint>".
+
+        This setter method stores the inputs and the type hints in two separate dictionaries:
+
+        * `inputs` - a dictionary of parameter names as keys and paths as values.
+        * `inputs_type_hints` - a dictionary of parameter names as keys and their type hints as values. If a parameter
+          is not in the type hints dictionary, its type hint will be `mlrun.DataItem` by default.
+
+        :param inputs: The inputs to set.
+        """
+        # Check if None, then set and return:
+        if inputs is None:
+            self._inputs = None
+            return
+
+        # Verify it's a dictionary:
+        inputs = self._verify_dict(inputs, "inputs")
+
+        # Prepare dictionaries to hold the cleared inputs and type hints:
+        cleared_inputs = {}
+        inputs_type_hints = {}
+
+        # Clear the current parsing configurations dictionary:
+        self._inputs_type_hints.clear()
+
+        # Clear the inputs from parsing configurations:
+        for input_key, input_value in inputs.items():
+            # Look for type hinted in input key:
+            if ":" in input_key:
+                # Separate the user input by colon:
+                input_key, input_type = self._separate_by_colon(value=input_key)
+                # Collect the type hint:
+                inputs_type_hints[input_key] = input_type
+            # Collect the cleared input key:
+            cleared_inputs[input_key] = input_value
+
+        # Set the now configuration free inputs:
+        self._inputs = cleared_inputs
+
+        # Set the collected input type hints:
+        self.inputs_type_hints = inputs_type_hints
+
+    @property
+    def inputs_type_hints(self) -> Dict[str, str]:
+        """
+        Get the input type hints. A dictionary of parameter names as keys and their type hints as values.
+
+        :return: The input type hints dictionary.
+        """
+        return self._inputs_type_hints
+
+    @inputs_type_hints.setter
+    def inputs_type_hints(self, inputs_type_hints: Dict[str, Union[type, str]]):
```

Comment:
I removed the option to let the user add type hints via `type` (meaning `pd.DataFrame` as type object rather than `"pandas.DataFrame"` as string) so this will be a regular setter from now on.